---
title: "MYH7 promoter motifs"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(biomartr)
library(biomaRt)
library(GenomicRanges)

require(RCurl) 
library(XML)
source("../util/_plot_theme.R")
```

## Retrieve MYH7 promoter motif matches from Ensembl

The Ensembl Regulatory Build seems to be one of the easiest / best resources for matched motifs in the human genome. I beleive that they also have information on bound proteins from ChIP-seq tracks -- which is awesome. Here we aim to do an analysis of the MYH7 promoter region to see which TFs could be linking MYH7b expression levels to MYH7 mRNA expression. 

```{r import-annotation}
# Retrieve motif matches for MYH7
# Let's use the annotation we've been using
gtf <- rtracklayer::import("../../util/gencode.v33.annotation.gtf")
genes <- gtf[which(gtf$type == "gene")]

myh7 <- genes[which(genes$gene_name == "MYH7")]
myh7_promoter <- promoters(myh7, upstream = 3e3, downstream = 3e3)
```

So we'll just grab the promoter region +/-3kb for MYH7B from the motif feature Ensembl track using the biomart webservice. 

```{bash retreive-motif-matches}
cd data
wget -O result.txt 'http://www.ensembl.org/biomart/martservice?query=<?xml version="1.0" encoding="UTF-8"?> 
<!DOCTYPE Query>
<Query  virtualSchemaName = "default" formatter = "TSV" header = "0" uniqueRows = "0" count = "" datasetConfigVersion = "0.6" >
			
	<Dataset name = "hsapiens_motif_feature" interface = "default" >
		<Filter name = "chromosome_name" value = "14"/>
		<Filter name = "start" value = "23432661"/>
		<Filter name = "end" value = "23438660"/>
		<Attribute name = "binding_matrix_id" />
		<Attribute name = "chromosome_name" />
		<Attribute name = "chromosome_start" />
		<Attribute name = "chromosome_end" />
		<Attribute name = "score" />
		<Attribute name = "feature_type_name" />
		<Attribute name = "display_label" />
	</Dataset>
</Query>'
```

What does the double colon mean in the ensembl regulatory build?

Iâ€™m going to assume for now that the two TFs separated by double colons means that the TFs both have a motif that corresponds to the match rather than that they need to be jointly bound.

```{r summarize-binding-sites}
binding_sites <- read.table("data/result.txt",
                            col.names = c("motif_id",
                                          "chr",
                                          "start",
                                          "end", 
                                          "score",
                                          "tf_name",
                                          "match_id"))

# Split the motifs that correspond to two transcription factors.
binding_sites <- binding_sites %>% 
  separate_rows("tf_name", sep = "::")



binding_site_summary <- binding_sites %>% group_by(tf_name) %>%
  summarize(count = n()) %>%
  arrange(-count) %>%
  mutate(tf_name = factor(tf_name, levels = tf_name))
```

Let's make a plot of the top binding sites at the promoter. 

```{r binding-site-count}
g <- ggplot(binding_site_summary %>% dplyr::slice(1:10), aes(x = tf_name, y = count))
g + geom_bar(stat = "identity") + 
  theme_paperwhite() +
  scale_y_continuous(expand = c(0,0)) +
  theme(panel.border = element_blank(),
        axis.line = element_line(colour = "black")) + 
  xlab("TF") +
  ylab("Num motif matches") +
  ggtitle("MYH7B promoter motif matches")
ggsave("figures/myh7b_promoter_motif_matches.png")
ggsave("figures/myh7b_promoter_motif_matches.pdf")
```


