---
title: 'Functional Analysis: lncMYH7b Knockdown RNA-seq in iPSCMs, Novogene'
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
editor_options:
  chunk_output_type: console
params:
  data_dir: !r file.path("../data", Sys.Date())
  dds_file: !r file.path("../data", Sys.Date(), "dds.rda")
  go_class: BP
  lfc_threshold: 0.5
  organism: Homo sapiens
  res_file: !r file.path("../data", Sys.Date(), "res_list_shrunken.rda")
  results_dir: !r file.path("../results", Sys.Date(), "TF_analysis")
bibliography: bibliography.bib
---
```{r libraries, echo=FALSE, message=FALSE, warning=FALSE}
# BiocManager::install("RcisTarget")


featherURL <- "https://resources-mirror.aertslab.org/cistarget/databases/homo_sapiens/hg38/refseq_r80/mc9nr/gene_based/hg38__refseq-r80__10kb_up_and_down_tss.mc9nr.feather"
# wget --no-check-certificate https://resources-mirror.aertslab.org/cistarget/databases/homo_sapiens/hg38/refseq_r80/mc9nr/gene_based/hg38__refseq-r80__10kb_up_and_down_tss.mc9nr.feather
# Old version
# featherURL <- "https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg19/refseq_r45/mc9nr/gene_based/hg19-tss-centered-10kb-7species.mc9nr.feather" 
# download.file(featherURL, destfile=basename(featherURL))


# BiocManager::install("RcisTarget.hg19.motifDBs.cisbpOnly.500bp")
suppressPackageStartupMessages({
library(RcisTarget)
# library(RcisTarget.hg38) 
library(DT)
library(data.table)
library(tidyverse)
library(zoo)
#require(visNetwork)
})
```

If you use RcisTarget in your research, please cite: 
#```{r citation, echo=FALSE}
print(citation("RcisTarget")[1], style="textVersion")
#```

# Overview of the workflow
The function `cisTarget()` allows to perform the motif-enrichment analysis on a gene list. The main input parameters are the gene list and the motif databases, which should be chosen depending on the organism and the search space around the TSS of the genes.
This is a sample on how to run the analysis (see the following sections for details):
```{r overview, eval=FALSE}
library(RcisTarget)
# Load gene sets to analyze. e.g.:
asokd<- read_csv("../01_ipscm_differential_expression/results/res_shrunken_asokd.csv")
asokd_genes_down <- asokd %>% filter(padj < 0.01, log2FoldChange < 0)
asokd_genes_up <- asokd %>% filter(padj < 0.01, log2FoldChange > 0)

spoe <- read_csv("../01_ipscm_differential_expression/results/res_shrunken_spoe.csv")

spoe_genes_down <- spoe %>% filter(padj < 0.01, log2FoldChange < 0)
spoe_genes_up <- spoe %>% filter(padj < 0.01, log2FoldChange > 0)

# geneLists <- GSEABase::GeneSet(genes, setName="geneListName") # alternative
# Select motif database to use (i.e. organism and distance around TSS)
data("motifAnnotations_hgnc")
motifAnnotations_hgnc <- motifAnnotations_hgnc %>% as.data.table()
motifRankings <- importRankings("hg38__refseq-r80__10kb_up_and_down_tss.mc9nr.feather")

# Motif enrichment analysis:
asodk_down_motifs <- cisTarget(asokd_genes_down$geneName, motifRankings,
                               motifAnnot=motifAnnotations_hgnc)
write_csv(asodk_down_motifs, "results/asokd_down_motifs.csv")
asodk_up_motifs <- cisTarget(asokd_genes_up$geneName, motifRankings,
                               motifAnnot=motifAnnotations_hgnc)
write_csv(asodk_up_motifs, "results/asokd_up_motifs.csv")

spoe_down_motifs <- cisTarget(spoe_genes_down$geneName, motifRankings,
                               motifAnnot=motifAnnotations_hgnc)
write_csv(spoe_down_motifs, "results/spoe_down_motifs.csv")
spoe_up_motifs <- cisTarget(spoe_genes_up$geneName, motifRankings,
                               motifAnnot=motifAnnotations_hgnc)
write_csv(spoe_up_motifs, "results/spoe_up_motifs.csv")

```



## Output

The final output of RcisTarget is a `data.table` containing the information about the motif enrichment and its annotation organized in the following fields:

* geneSet: Name of the gene set
* motif: ID of the motif
* NES: Normalized enrichment score of the motif in the gene-set
* AUC: Area Under the Curve (used to calculate the NES)
* TFinDB: Indicates whether the *highlightedTFs* are included within the high confidence annotation (two asterisks) or low confidence annotation (one asterisk).
* TF_highConf: Transcription factors annotated to the motif according to 'motifAnnot_highConfCat'.
* TF_lowConf: Transcription factors annotated to the motif according to 'motifAnnot_lowConfCat'.
* enrichedGenes: Genes that are highly ranked for the given motif. 
* nErnGenes: Number of genes highly ranked
* rankAtMax: Ranking at the maximum enrichment, used to determine the number of enriched genes.


```{r output}
motifEnrichmentTable_wGenes_wLogo <- addLogo(motifEnrichmentTable_wGenes)
resultsSubset <- motifEnrichmentTable_wGenes_wLogo[1:10,]
library(DT)
datatable(resultsSubset[,-c("enrichedGenes", "TF_lowConf"), with=FALSE], 
          escape = FALSE, # To show the logo
          filter="top", options=list(pageLength=5))
```


# Follow up examples

## TFs annotated to the enriched motifs
Note that the TFs are provided based on the motif annotation. They can be used as a guide to select relevant motifs or to prioritize some TFs, but the motif annotation does not imply that all the TFs appearing in the table regulate the gene list.

```{r anotatedTfs, cache=TRUE}
anotatedTfs <- lapply(split(motifEnrichmentTable_wGenes$TF_highConf,
                            motifEnrichmentTable$geneSet),
                      function(x) {
                        genes <- gsub(" \\(.*\\). ", "; ", x, fixed=FALSE)
                        genesSplit <- unique(unlist(strsplit(genes, "; ")))
                        return(genesSplit)
                        })
                      
anotatedTfs$hypoxia
```

## Building a network

```{r network, cache=FALSE, eval=FALSE}
signifMotifNames <- motifEnrichmentTable$motif[1:3]
incidenceMatrix <- getSignificantGenes(geneLists$hypoxia, 
                                       motifRankings,
                                       signifRankingNames=signifMotifNames,
                                       plotCurve=TRUE, maxRank=5000-20, 
                                       genesFormat="incidMatrix",
                                       method="aprox")$incidMatrix
library(reshape2)
edges <- melt(incidenceMatrix)
edges <- edges[which(edges[,3]==1),1:2]
colnames(edges) <- c("from","to")
```

*Output not shown:*
```{r visNetwork, eval=FALSE}
library(visNetwork)
motifs <- unique(as.character(edges[,1]))
genes <- unique(as.character(edges[,2]))
nodes <- data.frame(id=c(motifs, genes),   
      label=c(motifs, genes),    
      title=c(motifs, genes), # tooltip 
      shape=c(rep("diamond", length(motifs)), rep("elypse", length(genes))),
      color=c(rep("purple", length(motifs)), rep("skyblue", length(genes))))
visNetwork(nodes, edges) %>% visOptions(highlightNearest = TRUE, 
                                        nodesIdSelection = TRUE)
```


# sessionInfo()
This is the output of `sessionInfo()` on the system on which this document was compiled:
```{r}
date()
sessionInfo()
```

