---
title: 'Differential Expression: lncMYH7b Knockdown RNA-seq in iPSCMs, Novogene'
editor_options:
  chunk_output_type: console
---

```{r setup, message=FALSE}
options(stringsAsFactors = FALSE)
library(data.table)
setDTthreads(threads = 12)

library(rcartocolor) 
library(genefilter)
library(tximport)
library(tidyverse)
library(DESeq2)
library(pheatmap)
library(grDevices)
library(ggrepel)
library(venneuler)
library(ggforce)
library(ggpubr)

source("../../util/_util.R")
source("../../util/_plotting.R")
source("../util/_plot_theme.R")
```


```{r import_counts, message=FALSE, warning=FALSE, results='hide'}
# Read in gencode annotations
gtf <- rtracklayer::import("../../util/gencode.v33.annotation.gtf")

# Renaming for compatibility with tximport
# TODO: Do we still need to do this?
names(gtf@elementMetadata@listData)[which(names(gtf@elementMetadata@listData) == "gene_id")] <-
  "geneID"
names(gtf@elementMetadata@listData)[which(names(gtf@elementMetadata@listData) == "gene_name")] <-
  "geneName"
names(gtf@elementMetadata@listData)[which(names(gtf@elementMetadata@listData) == "gene_type")] <-
  "geneBiotype"
gtf@elementMetadata$id <- gtf@elementMetadata$geneID

genes <- gtf[which(gtf$type == "gene")]
gtfdf <- gtf %>% as.data.frame()
g2s <- as.data.frame(gtf@elementMetadata@listData) %>% 
  dplyr::select(geneID, geneName) %>% 
  distinct()
tx2gene <- gtfdf %>%
  mutate(TXNAME = transcript_id, GENEID = geneID) %>%
  dplyr::select(TXNAME, GENEID)
```



```{r}
# Read in sample annotations
samples <- read.csv("../../samplesheet.csv") %>%
  filter(sample_type == "hiPSCM")

samples_spoe <- samples %>% 
  filter(grepl("SP", condition))
samples_asokd <- samples %>%
  filter(grepl("ASO", condition))
rownames(samples_spoe) <- samples_spoe$sample_id
rownames(samples_asokd) <- samples_asokd$sample_id

# Factorize condition variables
samples_spoe$condition <- factor(samples_spoe$condition,
                                 levels = c("SP_CTL", "SP_OE"))
samples_asokd$condition <- factor(samples_asokd$condition,
                                 levels = c("ASO_CTL", "ASO_KD"))

# Read in the salmon counts data.
files <- file.path("../../ipscm_rnaseq/results/salmon", samples_spoe$sample_id, "quant.sf")
names(files) <- samples_spoe$sample_id
txisalmon_spoe <- tximport(files, type = "salmon", tx2gene = tx2gene)

files <- file.path("../../ipscm_rnaseq/results/salmon", samples_asokd$sample_id, "quant.sf")
names(files) <- samples_asokd$sample_id
txisalmon_asokd <- tximport(files, type = "salmon", tx2gene = tx2gene)
```

```{r retreive_tpm}
# TPM
tpm_spoe <- txisalmon_spoe$abundance %>%
  as.data.frame() %>% 
  rownames_to_column("geneID") %>%
  merge(g2s) %>%
  dplyr::select(geneID, geneName, everything())
tpm_asokd <- txisalmon_asokd$abundance %>%
  as.data.frame() %>% 
  rownames_to_column("geneID") %>%
  merge(g2s) %>%
  dplyr::select(geneID, geneName, everything())

# Export count tables
write_csv(tpm_spoe, "../../ipscm_rnaseq/results/tpm_spoe.csv")
write_csv(tpm_asokd, "../../ipscm_rnaseq/results/tpm_asokd.csv")
```


```{r create_DESeqDataSet}
# Create DESeq2 experiment objects.
samples_spoe <- samples_spoe[colnames(txisalmon_spoe$counts),]
stopifnot(all(rownames(samples_spoe) == colnames(txisalmon_spoe$counts)))

samples_asokd <- samples_asokd[colnames(txisalmon_asokd$counts),]
stopifnot(all(rownames(samples_asokd) == colnames(txisalmon_asokd$counts)))

# Let's also make sure the row ranges are in the same order
names(gtf) <- gtf$geneID
# Temporary fix for exon7skipped being created from an older gencode version.
# rownames(txisalmon_spoe$counts)[which(!(rownames(txisalmon_spoe$counts) %in% gtf$geneID))] <- "ENSG00000078814.16"

gtf <- gtf[rownames(txisalmon_spoe$counts)]
dds_spoe <- DESeqDataSetFromTximport(txisalmon_spoe,
                                        colData   = samples_spoe,
                                        design    = ~ condition,
                                        rowRanges = gtf)
gtf <- gtf[rownames(txisalmon_asokd$counts)]
dds_asokd <- DESeqDataSetFromTximport(txisalmon_asokd,
                                        colData   = samples_asokd,
                                        design    = ~ condition,
                                        rowRanges = gtf)
```


```{r prefilter}
keep <- rowSums(counts(dds_spoe)) >= 10
dds_spoe <- dds_spoe[keep, ]
keep <- rowSums(counts(dds_asokd)) >= 10
dds_asokd <- dds_asokd[keep, ]
```


```{r deseq}
dds_spoe <- DESeq(dds_spoe)
dds_asokd <- DESeq(dds_asokd)

save(dds_spoe, file = "results/dds_spoe.Rdata")
save(dds_asokd, file = "results/dds_asokd.Rdata")
```

```{r res_unshrunken}
res_spoe <- results(dds_spoe) 
res_asokd <- results(dds_asokd) 

# Export
resdf_spoe <- res_spoe %>% 
  as.data.frame() %>%
  rownames_to_column("geneID") %>%
  merge(g2s) %>%
  dplyr::select(geneID, geneName, everything())

resdf_asokd <- res_asokd %>% 
  as.data.frame() %>%
  rownames_to_column("geneID") %>%
  merge(g2s) %>%
  dplyr::select(geneID, geneName, everything())


# Just note that when the rmarkdown is knitted, it will consider
# The directory that it is saved in to be the working directory.
# For that reason, I usually set the working directory to be the 
# directory of the script that I'm currently working on.
# That way the knitting just works.
write_csv(resdf_spoe, "results/res_unshrunken_spoe.csv")
write_csv(resdf_asokd, "results/res_unshrunken_asokd.csv")
```

```{r res_shrunken}
# Use this to retrieve the coefficient.
# resultsNames(dds_spoe)
res_shrunken_spoe <- lfcShrink(dds_spoe, coef = "condition_SP_OE_vs_SP_CTL")
res_shrunken_asokd <- lfcShrink(dds_asokd, coef = "condition_ASO_KD_vs_ASO_CTL")

# Export
resdf_spoe <- res_shrunken_spoe %>% 
  as.data.frame() %>%
  rownames_to_column("geneID") %>%
  merge(g2s)

resdf_asokd <- res_shrunken_asokd %>% as.data.frame() %>%
  rownames_to_column("geneID") %>%
  merge(g2s)

write_csv(resdf_spoe, "results/res_shrunken_spoe.csv")
write_csv(resdf_asokd, "results/res_shrunken_asokd.csv")
```


# Plots

## PCA

Principal Component Analysis (PCA) is a statistical technique used to identify global patterns in high-dimensional datasets. It is commonly used to explore the similarity of biological samples in RNA-seq datasets. To achieve this, gene expression values are transformed into Principal Components (PCs), a set of linearly uncorrelated features which represent the most relevant sources of variance in the data, and subsequently visualized using a scatter plot.

Each point represents an RNA-seq sample. Samples with similar gene expression profiles are closer in the three-dimensional space. If provided, sample groups are indicated using different colors, allowing for easier interpretation of the results.


```{r variance_stabilization}
rld_spoe <- varianceStabilizingTransformation(dds_spoe)
rld_asokd <- varianceStabilizingTransformation(dds_asokd)
```


```{r plot_pca, message=FALSE}
DESeq2::plotPCA(rld_spoe, intgroup = "condition")  +
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() + 
  ggtitle("SP OE Sample PCA")
ggsave("figures/spoe_pca.pdf")
ggsave("figures/spoe_pca.png")

DESeq2::plotPCA(rld_asokd, intgroup = "condition")  +
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() + 
  ggtitle("ASO KD Sample PCA")
ggsave("figures/asokd_pca.pdf")
ggsave("figures/asokd_pca.png")
```

## Heatmap

This plot shows only differentially expressed genes on a per-sample basis. We have scaled the data by row and used the `ward.D2` method for clustering [@Ward1963-xf].

```{r plot_deg_heatmap}
plot_deg_heatmap <- function(res, rld, samp_df, interesting_groups = "condition", 
                             padj_thresh = 0.01, lfc_thresh = 0, title,
                             ann_colors = list(condition = c(SP_CTL  = "#424242", SP_OE = "#a8404c"))) {
  deg <- res %>%
    as.data.frame() %>%
    rownames_to_column("geneID") %>%
    filter(padj < padj_thresh, abs(log2FoldChange) > lfc_thresh)
  
  rld_counts <- rld@assays$data[[1]]
  deg_rld_counts <- rld_counts[which(rownames(rld_counts) %in% deg$geneID),]
  rld_scaled <- t(scale(t(deg_rld_counts)))
  
  anndf <- samp_df %>% dplyr::select(interesting_groups)
  
  
  col_pal <- c(colorRampPalette(colors = c("#424242", "#ffffff"))(49),
               "#ffffff","#ffffff",
               colorRampPalette(colors = c("#ffffff", "#a8404c"))(49))
  
  levels(anndf$condition)
  htmap <- pheatmap(rld_scaled, color = col_pal,
                 show_rownames = F,
                 annotation_col = anndf,
                 annotation_colors = ann_colors,
                 main = title)
  return(htmap)
}

spoe_heatmap <- plot_deg_heatmap(res_spoe, rld_spoe, samp_df = samples_spoe, interesting_groups = "condition",
                 padj_thresh = 0.01, lfc_thresh = 0, title = "SP OE DEG")
pdf("figures/spoe_deg_heatmap.pdf"); show(spoe_heatmap); dev.off()
png("figures/spoe_deg_heatmap.png"); show(spoe_heatmap); dev.off()


asokd_heatmap <- plot_deg_heatmap(res_asokd, rld_asokd, samp_df = samples_asokd, interesting_groups = "condition",
                 padj_thresh = 0.01, lfc_thresh = 0, title = "ASO KD DEG",
                 list(condition = c(ASO_CTL  = "#424242", ASO_KD = "#a8404c")))
pdf("figures/asokd_deg_heatmap.pdf"); show(asokd_heatmap); dev.off()
png("figures/asokd_deg_heatmap.png"); show(asokd_heatmap); dev.off()
```


```{r}
# MA plots
g <- ggplot(resdf_spoe, aes(x = log10(baseMean), y = log2FoldChange, label = geneName))
g + geom_hline(yintercept = 0) +
  geom_point(data = resdf_spoe %>% filter(padj >= 0.01), color = "#424242", alpha = 0.4) + 
  geom_point(data = resdf_spoe %>% filter(padj < 0.01), color = "#a8404c", alpha = 0.8) + 
  geom_text_repel(data = resdf_spoe %>% filter(abs(log2FoldChange) > 2)) +
  theme_paperwhite() + 
  ggtitle("SP OE")
ggsave("figures/spoe_maplot.pdf")
ggsave("figures/spoe_maplot.png")

g <- ggplot(resdf_asokd, aes(x = log10(baseMean), y = log2FoldChange, label = geneName))
g + geom_hline(yintercept = 0) +
  geom_point(data = resdf_asokd %>% filter(padj >= 0.01), color = "#424242", alpha = 0.4) + 
  geom_point(data = resdf_asokd %>% filter(padj < 0.01), color = "#a8404c", alpha = 0.8) + 
  geom_text_repel(data = resdf_asokd %>% filter(abs(log2FoldChange) > 2)) +
  theme_paperwhite() + 
  ggtitle("ASO KD")
ggsave("figures/asokd_maplot.pdf")
ggsave("figures/asokd_maplot.png")
```


```{r}
# Volcano plots 
g <- ggplot(resdf_spoe, aes(x = log2FoldChange, y = -log10(padj), label = geneName))
g + geom_vline(xintercept = 0, lty = 2) +
  geom_point(data = resdf_spoe %>% filter(padj >= 0.01), color = "#424242", alpha = 0.4) + 
  geom_point(data = resdf_spoe %>% filter(padj < 0.01), color = "#a8404c", alpha = 0.8) + 
  geom_text_repel(data = resdf_spoe %>% filter(abs(log2FoldChange) > 2)) +
  theme_paperwhite() + 
  ggtitle("SP OE")
ggsave("figures/spoe_volcano.pdf")
ggsave("figures/spoe_volcano.png")

g <- ggplot(resdf_asokd, aes(x = log2FoldChange, y = -log10(padj), label = geneName))
g + geom_vline(xintercept = 0, lty = 2) +
  geom_point(data = resdf_asokd %>% filter(padj >= 0.01), color = "#424242", alpha = 0.4) + 
  geom_point(data = resdf_asokd %>% filter(padj < 0.01), color = "#a8404c", alpha = 0.8) + 
  geom_text_repel(data = resdf_asokd %>% filter(abs(log2FoldChange) > 2)) +
  theme_paperwhite() + 
  ggtitle("ASO KD")
ggsave("figures/asokd_volcano.pdf")
ggsave("figures/asokd_volcano.png")
```


## MYH7b plot

This is the gene we intended to knockdown, so let's just see what this gene looks like.

```{r MYH7B}
gene_name <- "MYH7B"
gene_id <- g2s[which(g2s$geneName == gene_name),"geneID"] 
gene_tpm_spoe <- tpm_spoe[which(tpm_spoe$geneID == gene_id), ] %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples_spoe)
  
g <- ggplot(gene_tpm_spoe, aes(x = condition, y = tpm, color = condition))
g + geom_point() +
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5) + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() +
  ggtitle(paste0("SP OE -- ", gene_name, " expression"))
ggsave(paste0("figures/spoe_", gene_name, "_tpm.pdf"))
ggsave(paste0("figures/spoe_", gene_name, "_tpm.png"))

gene_tpm_asokd <- tpm_asokd[which(tpm_asokd$geneID == gene_id), ] %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples_asokd)

g <- ggplot(gene_tpm_asokd, aes(x = condition, y = tpm, color = condition))
g + geom_point() +
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5) + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() +
  ggtitle(paste0("ASO KD -- ", gene_name, " expression"))
ggsave(paste0("figures/asokd_", gene_name, "_tpm.pdf"))
ggsave(paste0("figures/asokd_", gene_name, "_tpm.png"))
```

```{r MYH7}
gene_name <- "MYH7"
gene_id <- g2s[which(g2s$geneName == gene_name),"geneID"] 
gene_tpm_spoe <- tpm_spoe[which(tpm_spoe$geneID == gene_id), ] %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples_spoe)
  
g <- ggplot(gene_tpm_spoe, aes(x = condition, y = tpm, color = condition))
g + geom_point() +
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5) + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() +
  ggtitle(paste0("SP OE -- ", gene_name, " expression"))
ggsave(paste0("figures/spoe_", gene_name, "_tpm.pdf"))
ggsave(paste0("figures/spoe_", gene_name, "_tpm.png"))

gene_tpm_asokd <- tpm_asokd[which(tpm_asokd$geneID == gene_id), ] %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples_asokd)

g <- ggplot(gene_tpm_asokd, aes(x = condition, y = tpm, color = condition))
g + geom_point() +
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5) + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() +
  ggtitle(paste0("ASO KD -- ", gene_name, " expression"))
ggsave(paste0("figures/asokd_", gene_name, "_tpm.pdf"))
ggsave(paste0("figures/asokd_", gene_name, "_tpm.png"))
```

```{r MYH6}
gene_name <- "MYH6"
gene_id <- g2s[which(g2s$geneName == gene_name),"geneID"] 
gene_tpm_spoe <- tpm_spoe[which(tpm_spoe$geneID == gene_id), ] %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples_spoe)
  
g <- ggplot(gene_tpm_spoe, aes(x = condition, y = tpm, color = condition))
g + geom_point() +
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5) + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() +
  ggtitle(paste0("SP OE -- ", gene_name, " expression"))
ggsave(paste0("figures/spoe_", gene_name, "_tpm.pdf"))
ggsave(paste0("figures/spoe_", gene_name, "_tpm.png"))

gene_tpm_asokd <- tpm_asokd[which(tpm_asokd$geneID == gene_id), ] %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples_asokd)

g <- ggplot(gene_tpm_asokd, aes(x = condition, y = tpm, color = condition))
g + geom_point() +
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5) + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() +
  ggtitle(paste0("ASO KD -- ", gene_name, " expression"))
ggsave(paste0("figures/asokd_", gene_name, "_tpm.pdf"))
ggsave(paste0("figures/asokd_", gene_name, "_tpm.png"))
```

### Overlap between experiments


```{r venn}
# Let's look at the overlaps between these two experiments in terms of differentially expressed genes.
resdf_asokd$experiment <- "asokd"
resdf_spoe$experiment <- "spoe"
res_combined <- bind_rows(resdf_asokd, resdf_spoe)
sig_combined <- res_combined %>% filter(padj < 0.01)
sig_combined$deg <- 1
sig_gene_matrix <- sig_combined %>% dplyr::select(geneID, experiment, deg) %>%
  pivot_wider(names_from = experiment, values_from = deg, values_fill = list(deg = 0)) %>%
  column_to_rownames("geneID") %>%
  as.matrix()


y <- venneuler(sig_gene_matrix)
venn_df <- data.frame(y$centers, diameters = y$diameters, labels = y$labels, stringsAsFactors = FALSE) %>%
  mutate(r = diameters/2)

# Calculate numeric values and add to plot
spoe_deg <- length(which(sig_gene_matrix[,"spoe"] == 1))
asokd_deg <- length(which(sig_gene_matrix[,"asokd"] == 1))
shared_deg <- length(which(sig_gene_matrix[,"asokd"] == 1 & sig_gene_matrix[,"spoe"] == 1))
asokd_unique <- asokd_deg - shared_deg
spoe_unique <- spoe_deg - shared_deg

venn_info <- data.frame("name" = c("spoe_deg", "asokd_deg", "shared_deg", "asokd_unique", "spoe_unique"),
                        "value" = c(paste0("n = ", spoe_deg),
                                    paste0("n = ", asokd_deg),
                                    shared_deg,
                                    aso_kd_unique,
                                    spoe_unique),
                        x = c(0.1,  1, 0.46,  0.726, 0.274),
                        y = c(0.74, 0.78, 0.5,  0.45, 0.45))


# Fix labels
venn_df[which(venn_df$labels == "asokd"), "labels"] <- "ASO KD"
venn_df[which(venn_df$labels == "spoe"), "labels"] <- "SP OE"

g <- ggplot(venn_df)
g + geom_circle(aes(x0 = x, y0 = y, r = r, fill=labels), alpha = .6, size = 0.3, color = "#212121") +
        geom_text(aes(x = x, y = y, label = labels), size = 7, color = "#212121") +
  coord_fixed() +
  scale_fill_manual(values = c("#424242","#a8404c")) + 
  theme_no_axes() +
  theme(legend.position = "none") +
  geom_text(data = venn_info, aes(x = x, y = y, label = value), color = "#212121") + 
  ggtitle("DEG: SP OE vs ASO KD", subtitle = "padj < 0.01")
ggsave("figures/ipscm_venn.png")
ggsave("figures/ipscm_venn.pdf")
```

```{r}
# Fold change correlation between experiments
res_combined_wide <- res_combined %>%
  dplyr::select(geneID, geneName, log2FoldChange, padj, experiment) %>%
  pivot_wider(id_cols = c("geneID", "geneName"), names_from = experiment,
              values_from = c("log2FoldChange", "padj")) %>%
  mutate(sig_asokd = padj_asokd < 0.01,
         sig_spoe = padj_spoe < 0.01,
         sig_both = sig_asokd & sig_spoe)
# Label points
res_combined_wide$sig <- "NS"
res_combined_wide[which(res_combined_wide$sig_asokd), "sig"] <- "DE in ASO KD"
res_combined_wide[which(res_combined_wide$sig_spoe), "sig"] <- "DE in SP OE"
res_combined_wide[which(res_combined_wide$sig_both), "sig"] <- "DE in both"
res_combined_wide$sig <- factor(res_combined_wide$sig, levels = c("NS", "DE in ASO KD", "DE in SP OE", "DE in both"))


# NOTE: This is using shrunken log2FoldChanges from DESeq2.
g <- ggplot(res_combined_wide, aes(x = log2FoldChange_asokd, y = log2FoldChange_spoe, color = sig))
g + geom_hline(color = "#212121", yintercept = 0, lty = 2) + 
  geom_vline(color = "#212121", xintercept = 0, lty = 2) +
  scale_color_manual(values = c("#424242","#71969F","#8B1D3B", "#DD1822"), name = "padj < 0.01") +
  geom_point(alpha = 0.7) + 
  theme_paperwhite() + 
  stat_cor() + 
  ggtitle("IPSCM: ASO KD vs SP OE")
ggsave("figures/ipscm_scatter.png")
ggsave("figures/ipscm_scatter.pdf")

# Let's export this handy table
write_csv(res_combined_wide, "results/ipscm_shrunken_lfc_compared.csv")
```

