---
title: 'Differential Expression: lncMYH7b Knockdown RNA-seq in iPSCMs, Novogene'
editor_options:
  chunk_output_type: console
---

```{r setup, message=FALSE}
options(stringsAsFactors = FALSE)
library(data.table)
setDTthreads(threads = 12)

library(rcartocolor) 
library(genefilter)
library(tximport)
library(tidyverse)
library(DESeq2)
library(pheatmap)
library(grDevices)

source("../../util/_util.R")
source("../../util/_plotting.R")
source("../util/_plot_theme.R")
```


```{r import_counts, message=FALSE, warning=FALSE, results='hide'}
# Read in gencode annotations
gtf <- rtracklayer::import("../../util/gencode.v33.annotation.gtf")
gtf <- gtf[which(gtf$type == "gene")]

#TODO: put this parsing in a setup file to clean this up.
names(gtf@elementMetadata@listData)[which(names(gtf@elementMetadata@listData) == "gene_id")] <-
  "geneID"
names(gtf@elementMetadata@listData)[which(names(gtf@elementMetadata@listData) == "gene_name")] <-
  "geneName"
names(gtf@elementMetadata@listData)[which(names(gtf@elementMetadata@listData) == "gene_type")] <-
  "geneBiotype"
gtf@elementMetadata$id <- gtf@elementMetadata$geneID

# Oh wow, okay. So, we made the exon7skipped version from v29
# and the gene id for MYH7b changed in the mean time. - corrected the ID (lindseyjoe)
# TODO: clean this up and re-run.
which(gtf$geneID == "ENSG00000078814.15")

gtfdf <- gtf %>% as.data.frame()
g2s <- as.data.frame(gtf@elementMetadata@listData) %>% 
  dplyr::select(geneID, geneName) %>% 
  distinct()
tx2gene <- gtfdf %>%
  mutate(TXNAME = transcript_id, GENEID = geneID) %>%
  dplyr::select(TXNAME, GENEID)

```



```{r}
# Read in sample annotations
samples <- read.csv("../../samplesheet.csv") %>%
  filter(sample_type == "hiPSCM")

samples_spoe <- samples %>% 
  filter(grepl("SP", condition))
samples_asokd <- samples %>%
  filter(grepl("ASO", condition))
rownames(samples_spoe) <- samples_spoe$sample_id
rownames(samples_asokd) <- samples_asokd$sample_id

# Factorize condition variables
samples_spoe$condition <- factor(samples_spoe$condition,
                                 levels = c("SP_CTL", "SP_OE"))
samples_asokd$condition <- factor(samples_asokd$condition,
                                 levels = c("ASO_CTL", "ASO_KD"))

# Read in the salmon counts data.
files <- file.path("../../ipscm_rnaseq/results/salmon", samples_spoe$sample_id, "quant.sf")
names(files) <- samples_spoe$sample_id
txisalmon_spoe <- tximport(files, type = "salmon", tx2gene = tx2gene)

files <- file.path("../../ipscm_rnaseq/results/salmon", samples_asokd$sample_id, "quant.sf")
names(files) <- samples_asokd$sample_id
txisalmon_asokd <- tximport(files, type = "salmon", tx2gene = tx2gene)
```

```{r retreive_tpm}
# TPM
tpm_spoe <- txisalmon_spoe$abundance %>%
  as.data.frame() %>% 
  rownames_to_column("geneID") %>%
  merge(g2s) %>%
  dplyr::select(geneID, geneName, everything())
tpm_asokd <- txisalmon_asokd$abundance %>%
  as.data.frame() %>% 
  rownames_to_column("geneID") %>%
  merge(g2s) %>%
  dplyr::select(geneID, geneName, everything())

# Export count tables
write_csv(tpm_spoe, "results/tpm_spoe.csv")
write_csv(tpm_asokd, "results/tpm_asokd.csv")
```


```{r create_DESeqDataSet}
# Create DESeq2 experiment objects.
samples_spoe <- samples_spoe[colnames(txisalmon_spoe$counts),]
stopifnot(all(rownames(samples_spoe) == colnames(txisalmon_spoe$counts)))

samples_asokd <- samples_asokd[colnames(txisalmon_asokd$counts),]
stopifnot(all(rownames(samples_asokd) == colnames(txisalmon_asokd$counts)))

# Let's also make sure the row ranges are in the same order
names(gtf) <- gtf$geneID
# Temporary fix for exon7skipped being created from an older gencode version.
# rownames(txisalmon_spoe$counts)[which(!(rownames(txisalmon_spoe$counts) %in% gtf$geneID))] <- "ENSG00000078814.16"

gtf <- gtf[rownames(txisalmon_spoe$counts)]
dds_spoe <- DESeqDataSetFromTximport(txisalmon_spoe,
                                        colData   = samples_spoe,
                                        design    = ~ condition,
                                        rowRanges = gtf)
gtf <- gtf[rownames(txisalmon_asokd$counts)]
dds_asokd <- DESeqDataSetFromTximport(txisalmon_asokd,
                                        colData   = samples_asokd,
                                        design    = ~ condition,
                                        rowRanges = gtf)
```


```{r prefilter}
keep <- rowSums(counts(dds_spoe)) >= 10
dds_spoe <- dds_spoe[keep, ]
keep <- rowSums(counts(dds_asokd)) >= 10
dds_asokd <- dds_asokd[keep, ]
```


```{r deseq}
dds_spoe <- DESeq(dds_spoe)
dds_asokd <- DESeq(dds_asokd)
```

```{r res_unshrunken}
res_spoe <- results(dds_spoe) 
res_asokd <- results(dds_asokd) 

# Export
resdf_spoe <- res_spoe %>% 
  as.data.frame() %>%
  rownames_to_column("geneID") %>%
  merge(g2s) %>%
  dplyr::select(geneID, geneName, everything())

resdf_asokd <- res_asokd %>% 
  as.data.frame() %>%
  rownames_to_column("geneID") %>%
  merge(g2s) %>%
  dplyr::select(geneID, geneName, everything())

write_csv(resdf_spoe, "results/res_unshrunken_spoe.csv")
write_csv(resdf_asokd, "results/res_unshrunken_asokd.csv")
```

```{r res_shrunken}
# Use this to retrieve the coefficient.
# resultsNames(dds_spoe)
res_shrunken_spoe <- lfcShrink(dds_spoe, coef = "condition_SP_OE_vs_SP_CTL")
res_shrunken_asokd <- lfcShrink(dds_asokd, coef = "condition_ASO_KD_vs_ASO_CTL")

# Export
resdf_spoe <- res_shrunken_spoe %>% 
  as.data.frame() %>%
  rownames_to_column("geneID") %>%
  merge(g2s)

resdf_asokd <- res_shrunken_asokd %>% as.data.frame() %>%
  rownames_to_column("geneID") %>%
  merge(g2s)

write_csv(resdf_spoe, "results/res_shrunken_spoe.csv")
write_csv(resdf_asokd, "results/res_shrunken_asokd.csv")
```


# Plots

## PCA

Principal Component Analysis (PCA) is a statistical technique used to identify global patterns in high-dimensional datasets. It is commonly used to explore the similarity of biological samples in RNA-seq datasets. To achieve this, gene expression values are transformed into Principal Components (PCs), a set of linearly uncorrelated features which represent the most relevant sources of variance in the data, and subsequently visualized using a scatter plot.

Each point represents an RNA-seq sample. Samples with similar gene expression profiles are closer in the three-dimensional space. If provided, sample groups are indicated using different colors, allowing for easier interpretation of the results.


```{r variance_stabilization}
rld_spoe <- varianceStabilizingTransformation(dds_spoe)
interestingGroups(rld_spoe) <- "condition"

rld_asokd <- varianceStabilizingTransformation(dds_asokd)
interestingGroups(rld_asokd) <- "condition"
```


```{r plot_pca, message=FALSE}
DESeq2::plotPCA(rld_spoe, intgroup = "condition")  +
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() + 
  ggtitle("SP OE Sample PCA")

DESeq2::plotPCA(rld_asokd, intgroup = "condition")  +
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() + 
  ggtitle("ASO KD Sample PCA")
```

## Heatmap

This plot shows only differentially expressed genes on a per-sample basis. We have scaled the data by row and used the `ward.D2` method for clustering [@Ward1963-xf].

```{r plot_deg_heatmap}
plot_deg_heatmap <- function(res, rld, samp_df, interesting_groups = "condition", 
                             padj_thresh = 0.01, lfc_thresh = 0, title,
                             ann_colors = list(condition = c(SP_CTL  = "#424242", SP_OE = "#a8404c"))) {
  deg <- res %>%
    as.data.frame() %>%
    rownames_to_column("geneID") %>%
    filter(padj < padj_thresh, abs(log2FoldChange) > lfc_thresh)
  
  rld_counts <- rld@assays$data[[1]]
  deg_rld_counts <- rld_counts[which(rownames(rld_counts) %in% deg$geneID),]
  rld_scaled <- t(scale(t(deg_rld_counts)))
  
  anndf <- samp_df %>% dplyr::select(interesting_groups)
  
  
  col_pal <- c(colorRampPalette(colors = c("#424242", "#ffffff"))(49),
               "#ffffff","#ffffff",
               colorRampPalette(colors = c("#ffffff", "#a8404c"))(49))
  
  levels(anndf$condition)
  show(pheatmap(rld_scaled, color = col_pal,
                 show_rownames = F,
                 annotation_col = anndf,
                 annotation_colors = ann_colors,
                 main = title))
  
}

plot_deg_heatmap(res_spoe, rld_spoe, samp_df = samples_spoe, interesting_groups = "condition",
                 padj_thresh = 0.01, lfc_thresh = 0, title = "SP OE DEG")

plot_deg_heatmap(res_asokd, rld_asokd, samp_df = samples_asokd, interesting_groups = "condition",
                 padj_thresh = 0.01, lfc_thresh = 0, title = "ASO KD DEG",
                 list(condition = c(ASO_CTL  = "#424242", ASO_KD = "#a8404c")))
```


```{r}
# MA plots
library(ggrepel)
g <- ggplot(resdf_spoe, aes(x = log10(baseMean), y = log2FoldChange, label = geneName))
g + geom_hline(yintercept = 0) +
  geom_point(data = resdf_spoe %>% filter(padj >= 0.01), color = "#424242", alpha = 0.4) + 
  geom_point(data = resdf_spoe %>% filter(padj < 0.01), color = "#a8404c", alpha = 0.8) + 
  geom_text_repel(data = resdf_spoe %>% filter(abs(log2FoldChange) > 2)) +
  theme_paperwhite() + 
  ggtitle("SP OE")

g <- ggplot(resdf_asokd, aes(x = log10(baseMean), y = log2FoldChange, label = geneName))
g + geom_hline(yintercept = 0) +
  geom_point(data = resdf_asokd %>% filter(padj >= 0.01), color = "#424242", alpha = 0.4) + 
  geom_point(data = resdf_asokd %>% filter(padj < 0.01), color = "#a8404c", alpha = 0.8) + 
  geom_text_repel(data = resdf_asokd %>% filter(abs(log2FoldChange) > 2)) +
  theme_paperwhite() + 
  ggtitle("ASO KD")
```


```{r}
# Volcano plots 
g <- ggplot(resdf_spoe, aes(x = log2FoldChange, y = -log10(padj), label = geneName))
g + geom_vline(xintercept = 0, lty = 2) +
  geom_point(data = resdf_spoe %>% filter(padj >= 0.01), color = "#424242", alpha = 0.4) + 
  geom_point(data = resdf_spoe %>% filter(padj < 0.01), color = "#a8404c", alpha = 0.8) + 
  geom_text_repel(data = resdf_spoe %>% filter(abs(log2FoldChange) > 2)) +
  theme_paperwhite() + 
  ggtitle("SP OE")

g <- ggplot(resdf_asokd, aes(x = log2FoldChange, y = -log10(padj), label = geneName))
g + geom_vline(xintercept = 0, lty = 2) +
  geom_point(data = resdf_asokd %>% filter(padj >= 0.01), color = "#424242", alpha = 0.4) + 
  geom_point(data = resdf_asokd %>% filter(padj < 0.01), color = "#a8404c", alpha = 0.8) + 
  geom_text_repel(data = resdf_asokd %>% filter(abs(log2FoldChange) > 2)) +
  theme_paperwhite() + 
  ggtitle("ASO KD")
```


## MYH7b plot

This is the gene we intended to knockdown, so let's just see what this gene looks like.

```{r myh7b}

# Gene id for MYH7B
gene_id <- "ENSG00000078814.16"
gene_tpm_spoe <- tpm_spoe[which(tpm_spoe$geneID == gene_id), ] %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples_spoe)
  
g <- ggplot(gene_tpm_spoe, aes(x = condition, y = tpm, color = condition))
g + geom_point() +
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5) + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() +
  ggtitle("SP OE -- MYH7B expression")


gene_tpm_asokd <- tpm_asokd[which(tpm_asokd$geneID == gene_id), ] %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples_asokd)

g <- ggplot(gene_tpm_asokd, aes(x = condition, y = tpm, color = condition))
g + geom_point() +
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5) + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() +
  ggtitle("ASO KD -- MYH7B expression")
```


```{r myh7}

gene_id <- "ENSG00000092054.13"
gene_tpm_spoe <- tpm_spoe[which(tpm_spoe$geneID == gene_id), ] %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples_spoe)
  
g <- ggplot(gene_tpm_spoe, aes(x = condition, y = tpm, color = condition))
g + geom_point() +
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5) + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() +
  ggtitle("SP OE -- MYH7 expression")


gene_tpm_asokd <- tpm_asokd[which(tpm_asokd$geneID == gene_id), ] %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples_asokd)

g <- ggplot(gene_tpm_asokd, aes(x = condition, y = tpm, color = condition))
g + geom_point() +
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5) + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() +
  ggtitle("ASO KD -- MYH7 expression")

```


```{r myh6}
gene_name <- "ENSG00000197616.12"
gene_tpm_spoe <- tpm_spoe[which(tpm_spoe$geneID == gene_id), ] %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples_spoe)
  
g <- ggplot(gene_tpm_spoe, aes(x = condition, y = tpm, color = condition))
g + geom_point() +
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5) + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() +
  ggtitle("SP OE -- MYH6 expression")


gene_tpm_asokd <- tpm_asokd[which(tpm_asokd$geneID == gene_id), ] %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples_asokd)

g <- ggplot(gene_tpm_asokd, aes(x = condition, y = tpm, color = condition))
g + geom_point() +
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5) + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  theme_paperwhite() +
  ggtitle("ASO KD -- MYH6 expression")
```
