---
title: 'Differential Expression: lncMYH7b Knockdown RNA-seq in iPSCMs, Novogene'
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
params:
  alpha: 0.01
  contrasts: !r list( c( factor = "lncMYH7b", numerator = "ASO_KD", denominator = "ASO_CTL"))
  data_dir: !r file.path("../data", Sys.Date())
  design: !r formula("~ lncMYH7b")
  dropbox_dir: lncMYH7B_KD
  dropbox_token: ~/dropbox.RDS
  lfc_threshold: 0
  results_dir: !r file.path("../results", Sys.Date(), "differential_expression")
editor_options:
  chunk_output_type: console
---

```{r setup, message=FALSE}
options(stringsAsFactors = FALSE)
bcbioRNASeq::prepareRNASeqTemplate()
source("_setup.R"); source("../../util/_util.R"); source("../../util/_plotting.R")

library(data.table); library(rcartocolor); library(genefilter); library(tximport)
```

## Introduction


```{r import_counts, message=FALSE, warning=FALSE, results='hide'}

# Read in gencode annotations
gtf <- rtracklayer::import("../../util/gencode.v33.annotation.gtf")


#TODO: put this parsing in a setup file to clean this up.
names(gtf@elementMetadata@listData)[which(names(gtf@elementMetadata@listData) == "gene_id")] <-
  "geneID"
names(gtf@elementMetadata@listData)[which(names(gtf@elementMetadata@listData) == "gene_name")] <-
  "geneName"
names(gtf@elementMetadata@listData)[which(names(gtf@elementMetadata@listData) == "gene_type")] <-
  "geneBiotype"
gtf@elementMetadata$id <- gtf@elementMetadata$geneID

# Oh wow, okay. So, we made the exon7skipped version from v29
# and the gene id for MYH7b changed in the mean time. 
# TODO: clean this up and re-run.
which(gtf$geneID == "ENSG00000078814.15")

gtfdf <- gtf %>% as.data.frame()
g2s <- as.data.frame(gtf@elementMetadata@listData) %>% dplyr::select(geneID, geneName) %>% distinct()
tx2gene <- gtfdf %>%
  mutate(TXNAME = transcript_id, GENEID = geneID) %>%
  dplyr::select(TXNAME, GENEID)

```

TODO: is this still relevant?
Sample 12047_3 has very low read counts and we will drop it for the rest of the analysis.

```{r drop_sample}


# Read in sample annotations
samples <- read.csv("../../samplesheet.csv") %>%
  filter(sample_type == "hiPSCM")
samples_spoe <- samples %>% 
  filter(grepl("SP", condition))
samples_asokd <- samples %>%
  filter(grepl("ASO", condition))
rownames(samples_spoe) <- samples_spoe$sample_id
rownames(samples_asokd) <- samples_asokd$sample_id

# Factorize condition variables
samples_spoe$condition <- factor(samples_spoe$condition,
                                 levels = c("SP_CTL", "SP_OE"))
samples_asokd$condition <- factor(samples_asokd$condition,
                                 levels = c("ASO_CTL", "ASO_KD"))

# Read in the salmon counts data.
files <- list.files("../../ipscm_rnaseq/results/salmon")


files <- file.path("../../ipscm_rnaseq/results/salmon", samples_spoe$sample_id, "quant.sf")
names(files) <- samples_spoe$sample_id
txisalmon_spoe <- tximport(files, type = "salmon", tx2gene = tx2gene)

files <- file.path("../../ipscm_rnaseq/results/salmon", samples_asokd$sample_id, "quant.sf")
names(files) <- samples_asokd$sample_id
txisalmon_asokd <- tximport(files, type = "salmon", tx2gene = tx2gene)

# gene_ids_used <- rownames(txisalmon_spoe$counts)
# gtf <- gtf[which(gtf$geneID %in% gene_ids_used)]
gtf <- gtf[which(gtf$type == "gene")]
```

```{r convert_counts}
# TPM
tpm_spoe <- txisalmon_spoe$abundance
tpm_asokd <- txisalmon_asokd$abundance

# Export count tables
write.table(tpm_spoe, file = "results/tpm_spoe.txt", 
            sep="\t", row.names=T, col.names=T, quote=F)
write.table(tpm_asokd, file = "results/tpm_asokd.txt", 
            sep="\t", row.names=T, col.names=T, quote=F)
```


```{r create_DESeqDataSet}
# Create DESeq2 experiment objects.
samples_spoe <- samples_spoe[colnames(txisalmon_spoe$counts),]
stopifnot(all(rownames(samples_spoe) == colnames(txisalmon_spoe$counts)))

samples_asokd <- samples_asokd[colnames(txisalmon_asokd$counts),]
stopifnot(all(rownames(samples_asokd) == colnames(txisalmon_asokd$counts)))

# Let's also make sure the row ranges are in the same order
names(gtf) <- gtf$geneID
# Temporary fix for exon7skipped being created from an older gencode version.
# rownames(txisalmon_spoe$counts)[which(!(rownames(txisalmon_spoe$counts) %in% gtf$geneID))] <- "ENSG00000078814.16"

gtf <- gtf[rownames(txisalmon_spoe$counts)]
dds_spoe <- DESeq2::DESeqDataSetFromTximport(txisalmon_spoe,
                                        colData   = samples_spoe,
                                        design    = ~ condition,
                                        rowRanges = gtf)
gtf <- gtf[rownames(txisalmon_asokd$counts)]
dds_asokd <- DESeq2::DESeqDataSetFromTximport(txisalmon_asokd,
                                        colData   = samples_asokd,
                                        design    = ~ condition,
                                        rowRanges = gtf)
```

# Pre-filtering

While it is not necessary to pre-filter low count genes before running the DESeq2 functions, there are two reasons which make pre-filtering useful: by removing rows in which there are very few reads, we reduce the memory size of the dds data object, and we increase the speed of the transformation and testing functions within DESeq2. Here we perform a minimal pre-filtering to keep only rows that have at least 10 reads total. Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the `results()` function.

```{r prefilter}
# Note that this criteria can be made more stringent
keep <- rowSums(counts(dds_spoe)) >= 10
dds_spoe <- dds_spoe[keep, ]
keep <- rowSums(counts(dds)) >= 10
dds_asokd <- dds_asokd[keep, ]

```


```{r deseq}
dds_spoe <- DESeq(dds_spoe)
dds_asokd <- DESeq(dds_asokd)
```


```{r variance_stabilization}
# Alternatively, can use `rlog()` here, but it is slower
rld_spoe <- varianceStabilizingTransformation(dds_spoe)
interestingGroups(rld_spoe) <- "condition"

rld_asokd <- varianceStabilizingTransformation(dds_asokd)
interestingGroups(rld_asokd) <- "condition"
```




```{r res_unshrunken}
res_spoe <- results(dds_spoe) 
res_asokd <- results(dds_asokd) 

```

Now let's calculate shrunken log2 fold change values with `DESeq2::lfcShrink()`. We're using the "normal" shrinkage estimator (default in DESeq2); the "apeglm" shrinkage estimator is also promising but doens't work well with complex contrast designs.

```{r res_shrunken}
# For `type` arguments other than "normal", `coef` argument is required.
# Otherwise can use `contrast`, which is more intuitive and readable.
# If using `coef` number, must match order in `resultsNames(dds)`.
# The new apeglm method is great but currently only works with coef.
# help("lfcShrink", "DESeq2")
# help("coef", "DESeq2")
# help("resultsNames", "DESeq2")
# This step can be a little slow and sped up with the `parallel` argument.

resultsNames(dds_asokd)
res_shrunken_spoe <- lfcShrink(dds_spoe, coef = "condition_SP_OE_vs_SP_CTL")
res_shrunken_asokd <- lfcShrink(dds_asokd, coef = "condition_ASO_KD_vs_ASO_CTL")
```

```{r}
# export

resdf_spoe <- res_shrunken_spoe %>% as.data.frame() %>%
  rownames_to_column("geneID") %>%
  merge(g2s)
write_csv(resdf_spoe, "results/res_shrunken_spoe.csv")
resdf_asokd <- res_shrunken_asokd %>% as.data.frame() %>%
  rownames_to_column("geneID") %>%
  merge(g2s)
write_csv(resdf_asokd, "results/res_shrunken_asokd.csv")
```




# Plots

## PCA

Principal Component Analysis (PCA) is a statistical technique used to identify global patterns in high-dimensional datasets. It is commonly used to explore the similarity of biological samples in RNA-seq datasets. To achieve this, gene expression values are transformed into Principal Components (PCs), a set of linearly uncorrelated features which represent the most relevant sources of variance in the data, and subsequently visualized using a scatter plot.

Each point represents an RNA-seq sample. Samples with similar gene expression profiles are closer in the three-dimensional space. If provided, sample groups are indicated using different colors, allowing for easier interpretation of the results.

```{r plot_deg_pca, message=FALSE}
plotDEGPCA(res, counts = rld, alpha = 0.01, lfcThreshold = 0)
```


## Heatmap

This plot shows only differentially expressed genes on a per-sample basis. We have scaled the data by row and used the `ward.D2` method for clustering [@Ward1963-xf].

```{r plot_deg_heatmap}
plotDEGHeatmap(res, counts = rld, clusteringMethod = "ward.D2", scale = "row", alpha = 0.01, 
               color = pals::kovesi.diverging_bwr_40_95_c42,
               legendColor = pals::watlington)

plotMA(res, lfcThreshold = params$lfc_threshold, ntop = 10, gene2symbol = g2s, alpha = 0.01)

exsk <- data.frame("geneID" = "ENSG00000078814.16.1", "geneName" = "MYH7B_exon7_skipped")
g2s <- bind_rows(g2s, exsk)
plotVolcano(res, lfcThreshold = params$lfc_threshold, ntop = 10, gene2symbol = g2s, alpha = 0.01,
            sigPointColor = c(upregulated = "#a8404c", downregulated = "#a8404c"))
```




## MYH7b plot

This is the gene we intended to knockdown, so let's just see what this gene looks like.

```{r myh7b}

"ENSG00000078814.15"

gene_name <- "ENSG00000078814.16"
gene_tpm <- tpm[which(row.names(tpm) == gene_name), ] %>%
    as.data.frame() %>% rownames_to_column()
  colnames(gene_tpm) <- c("SampleID", "tpm")
  gene_tpm <- merge(gene_tpm, samples)

g <- ggplot(gene_tpm, aes(x = SampleID, y = tpm, color = condition))
g + geom_point()  + ggtitle("MYH7B expression")
  



```


```{r myh7}

"ENSG00000092054.12"

gene_name <- "ENSG00000092054.12"
gene_tpm <- tpm[which(row.names(tpm) == gene_name), ] %>%
    as.data.frame() %>% rownames_to_column()
  colnames(gene_tpm) <- c("SampleID", "tpm")
  gene_tpm <- merge(gene_tpm, samples)

g <- ggplot(gene_tpm, aes(x = SampleID, y = tpm, color = condition))
g + geom_point()  + ggtitle("MYH7 expression")

```


```{r myh6}

"ENSG00000197616.11"

gene_name <- "ENSG00000197616.11"
gene_tpm <- tpm[which(row.names(tpm) == gene_name), ] %>%
    as.data.frame() %>% rownames_to_column()
  colnames(gene_tpm) <- c("SampleID", "tpm")
  gene_tpm <- merge(gene_tpm, samples)

g <- ggplot(gene_tpm, aes(x = SampleID, y = tpm, color = lncMYH7b))
g + geom_point()  + ggtitle("MYH6 expression")

```
# Results tables

Subset the results into separate tables, containing all genes, differentially expressed genes in both directions, and directional tables.

```{r results_tables, results="asis"}
# Here we're creating subset tables of the DEGs, and adding the normalized
# counts used by DESeq2 for the differential expression analysis.
mdout <- capture.output(res_tbl_list <- mapply(
    FUN = resultsTables,
    results = res_list_shrunken,
    MoreArgs = list(
        counts = dds,
        lfcThreshold = params$lfc,
        summary = TRUE,
        headerLevel = 2,
        write = TRUE,
        dir = params$results_dir
        # dropboxDir = params$dropbox_dir,
        # rdsToken = params$dropbox_token
    ),
    SIMPLIFY = FALSE,
    USE.NAMES = TRUE
))

# This is a little hack to force the dropbox links to autodownload.
mdout <- gsub("dl=0", "dl=1", mdout)
cat(paste(mdout,collapse="\n"))


saveData(res_tbl_list, dir = params$data_dir)
```



Differentially expressed gene (DEG) tables are sorted by BH-adjusted P value, and contain the following columns:

- `baseMean`: Mean of the normalized counts per gene for all samples.
- `log2FoldChange`: log2 fold change.
- `lfcSE`: log2 standard error.
- `stat`: Wald statistic.
- `pvalue`: Walt test *P* value.
- `padj`: BH adjusted Wald test *P* value (corrected for multiple comparisons; aka FDR).

## Top tables

Only the top up- and down-regulated genes (arranged by log2 fold change) are shown.

## Top tables

Only the top up- and down-regulated genes (arranged by log2 fold change) are shown.

```{r top_tables, results="asis"}
invisible(mapply(
    FUN = topTables,
    object = res_tbl_list,
    MoreArgs = list(coding = FALSE)
))
```

```{r footer, child="_footer.Rmd"}
```
