---
title: "MYH7-MYH8 ratio"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
source("../util/_util.R")
source("../util/_setup.R")
```

```{r}
# This version has the exon skipped transcript appended.
# gtf <- rtracklayer::import("../../gencode.v32.annotation.gtf")
# genes <- gtf[which(gtf$type == "gene")]
# rtracklayer::export(genes, "gencode.v32.annotation.genes.gtf")
# genes <- rtracklayer::import("gencode.v32.annotation.genes.gtf")
# 
# 
# counts <- read.table("../../results/featureCounts/merged_gene_counts.txt", header = T)
# 
# g2s <- counts[,c(1,2)]
# names(g2s) <- c("gene_id", "gene_name")
# 
# 
# # Trim the sample names.
# colnames(counts) <- sapply(colnames(counts), function(x) {
#   unlist(strsplit(x, "Aligned"))[[1]]
# })
# samples <- read.csv("../../samplesheet.csv")
# 
# samples$condition <- sapply(samples$sample_title, function(x) {
#   gsub('[[:digit:]]+', '', x)
# })
# 
# 
# # For some reason these samples doesn't exist
# # SRR7426830
# # SRR7426847
# samples <- samples %>% filter(run_accession != "SRR7426830")
# samples <- samples %>% filter(run_accession != "SRR7426847")
# 
# # Convert of tpm
# fc <- fCountReader("../../results/featureCounts/gene_counts/",
#                        samples$run_accession,
#                        "Aligned.sortedByCoord.out_gene.featureCounts.txt")
# tpm <- countConverter(fc, return = "TPM") %>%
#   as.data.frame() %>%
#   rownames_to_column(var = "gene_id") %>%
#   merge(g2s)
```

# MHY7 / MYH8 tpm plots

We may want to take the ratio of alpha to beta within each sample, maybe from the raw counts, maybe from some DEseq normalized count step, but for now, let's just look at the TPMs across groups.

```{r}
### MYH6 (alpha) main transcript: ENST00000405093.8
### MYH7 (beta) main transcript: ENST00000355349.4
### Exon skipped MYH7B transcript: ENST00000618182.5_exon_8_skipped

myh_genes <- tx2gene %>% filter(TXNAME %in% c("ENST00000405093.8", "ENST00000355349.4", "ENST00000618182.5_exon_8_skipped")) %>%
  distinct() %>%
  filter(!is.na(TXNAME))

myh <- tpm %>% filter(transcript_id %in% myh_genes$TXNAME) %>%
  gather(run_accession, tpm, 2:62)
names(myh_genes) <- c("transcript_id", "gene_id", "gene_name")

myh <- merge(myh, myh_genes)
names(myh)
names(samples)[1] <- "run_accession"
myh <- merge(myh, samples %>% dplyr::select(run_accession, condition))

myh_wide <- myh %>% 
  dplyr::select(run_accession, condition, gene_name, tpm) %>%
  pivot_wider(names_from = gene_name, values_from = tpm) %>%
  distinct()
table(myh_wide$condition)
g <- ggplot(myh_wide, aes(x = MYH7B, y = MYH7, color = condition))
g + geom_point() 


myh_wide$condition <- factor(myh_wide$condition, levels = c("NF", "DCM", "ICM"))

g <- ggscatter(myh_wide, x = "MYH7B", y = "MYH7",
          color = "condition",
          shape = "condition",
          add = "reg.line",   
          palette = c("#424242", "#752D2D","#C24A4A" ),
          conf.int = TRUE,                                
          add.params = list(color = "black",
                            fill = "lightgray"))+
  stat_cor(method = "pearson", size = 5) + 
  ylab("MHY7 tpm") + 
  xlab("MYH7B_exon7skipped tpm") + 
  ggtitle("lncMYH7B vs. MYH7 expression") +
  theme_pubr(base_size = 18)
show(g)

ggsave("figures/lncMYH7B_vs_MYH7_all_conditions.pdf", useDingbats = FALSE,
       height = 7, width = 7)

# Let's plot them separately now
g <- ggscatter(myh_wide %>% filter(condition == "NF"), x = "MYH7B", y = "MYH7",
          color = "#424242",
          add = "reg.line",   
          conf.int = TRUE,                                
          add.params = list(color = "black",
                            fill = "lightgray"))+
  stat_cor(method = "pearson", size = 5) + 
  ylab("MHY7 tpm") + 
  xlab("MYH7B_exon7skipped tpm") + 
  ggtitle("NF: lncMYH7B vs. MYH7 expression") +
  theme_pubr(base_size = 16)
show(g)

ggsave("figures/lncMYH7B_vs_MYH7_all_NF.pdf", useDingbats = FALSE,
       width = 7, height = 4)

g <- ggscatter(myh_wide %>% filter(condition == "DCM"), x = "MYH7B", y = "MYH7",
          color = "#752D2D",
          add = "reg.line",   
          conf.int = TRUE,                                
          add.params = list(color = "black",
                            fill = "lightgray"))+
  stat_cor(method = "pearson", size = 5) + 
  ylab("MHY7 tpm") + 
  xlab("MYH7B_exon7skipped tpm") + 
  ggtitle("DCM: lncMYH7B vs. MYH7 expression") +
  theme_pubr(base_size = 16)
show(g)
ggsave("figures/lncMYH7B_vs_MYH7_all_DCM.pdf", useDingbats = FALSE,
       width = 7, height = 4)

g <- ggscatter(myh_wide %>% filter(condition == "ICM"), x = "MYH7B", y = "MYH7",
          color = "#C24A4A",
          add = "reg.line",   
          conf.int = TRUE,                                
          add.params = list(color = "black",
                            fill = "lightgray"))+
  stat_cor(method = "pearson", size = 5) + 
  ylab("MHY7 tpm") + 
  xlab("MYH7B_exon7skipped tpm") + 
  ggtitle("ICM: lncMYH7B vs. MYH7 expression") +
  theme_pubr(base_size = 16)
show(g)

ggsave("figures/lncMYH7B_vs_MYH7_all_ICM.pdf", useDingbats = FALSE,
       width = 7, height = 4)
```

Let's take the ratio in each sample.

```{r}
myh_wide <- myh %>% 
select(-gene_id) %>%
  spread(gene_name, tpm)
myh_wide$total_tpm <- myh_wide$MYH6 + myh_wide$MYH7
myh_wide$alpha_fraction <- myh_wide$MYH6 / myh_wide$total_tpm
myh_wide$beta_fraction <- myh_wide$MYH7 / myh_wide$total_tpm


myh_ab <- myh_wide %>%
  select(run_accession, condition, alpha_fraction, beta_fraction) %>%
  gather(protein, fraction, 3:4)
myh_ab$protein <- gsub("_fraction", "", myh_ab$protein)
g <- ggplot(myh_ab, aes(x = run_accession, y = fraction, fill = protein))
g + geom_bar(position = "stack", stat = "identity") + 
  facet_wrap(~condition, scales = "free_x")
```


```{r}
# let's get the mean per condition
myh_mean_fraction <- myh_ab %>% group_by(condition, protein) %>%
  summarize(mean_fraction  = mean(fraction))
# pdf("alpha_beta_fraction_human_heart.pdf")
g <- ggplot(myh_mean_fraction, aes(x = condition, y = mean_fraction, fill = protein))
g + geom_bar(position = "stack", stat = "identity") + 
  scale_fill_manual(values = c("#a8404c", "#424242")) + 
  coord_flip() + 
  ggtitle(bquote(alpha ~ "/" ~ beta ~ " fraction"))
# dev.off()

# NF -- Non-failing
# DCM --- dilated cardiomyopathy
# ICM -- ischemic cardiomyopathy
```


```{r}
library(ggpubr)

g <- ggboxplot(myh_ab %>% filter(protein == "alpha"), 
               x = "condition", y = "fraction",
               color = "#424242",
               # palette =c("#00AFBB", "#E7B800", "#FC4E07"),
               add = "jitter")
# pdf("alpha_fraction_human_heart.pdf")
my_comparisons <- list( c("DCM", "ICM"),c("NF", "DCM"), c("NF", "ICM"))
g + stat_compare_means(comparisons = my_comparisons)+ # Add pairwise comparisons p-value
  stat_compare_means(label.y = 0.65) +
  xlab("") + 
  ylab(bquote(alpha ~ "-fraction")) + 
  ggtitle(bquote(alpha ~ "-fraction in human heart failure"))
# dev.off()

```

# Exon skipped fraction

```{r}
# Don't really need tximport at this point.
# library(tximport)
# 
# files <- file.path("../../results/salmon/", samples$run_accession, "quant.sf")
# names(files) <- samples$run_accession
# all(file.exists(files))

# 
# salmon_transcript_counts <- readRDS("../../results/salmon/salmon_merged_transcript_counts.rds")

salmon_transcript_tpm <- read.csv("results/human_heart_transcript_tpm.csv")

# tx2gene <- read_csv(file.path("../../human_heart_rnaseq/results/salmon", "tx2gene.csv"), col_names = FALSE)
# colnames(tx2gene) <- c("transcript_id", "gene_id", "gene_name")

gtfdf <- gtf %>% as.data.frame()
g2s <- as.data.frame(gtf@elementMetadata@listData) %>% 
  dplyr::select(geneID, geneName) %>% 
  distinct()
tx2gene <- gtfdf %>%
  mutate(TXNAME = transcript_id, GENEID = geneID) %>%
  dplyr::select(TXNAME, GENEID, geneName)

myh7b_trancripts <- tx2gene[which(tx2gene$geneName == "MYH7B"),] %>% distinct()

myh7b_quant <- salmon_transcript_tpm[salmon_transcript_tpm$transcript_id %in% myh7b_trancripts$TXNAME, ] 

myh7b_long <- myh7b_quant %>% gather(run_accession, tpm, 2:62) %>%
  merge(samples)

g <- ggplot(myh7b_long, aes(x = transcript_id, y = tpm))
g + geom_point()

# There is an extendo-transcript that starts from the same promoter as GSS
# but that isn't annotated in gencode, only in NCBI Refseq
# We don't know which isoform was used by the previous collaborator,
# however, his results were such that the exon skipped version
# of the transcript was more highly expressed.
# It is likely then that he used the ENST00000618182.4
# Which is also the version that we used for the exon skipping. 
# however, here the most highly expressed version appears to be
# ENST00000453028.5. Which has one more exon upstream.
# howver, on the browser, it doesn't appear there are cage peaks where
# that promoter would be....
```

Okay, perhaps to get arround the difficulty of determining which transcript to use to compare the exon-skipped transcript, we can just correlate the expression of exon7-skipped transcript to the ratio of alpha-beta.

```{r}

e7skipped <- myh7b_long[grep("skipped", myh7b_long$transcript_id),] %>%
  select(run_accession, tpm)
colnames(e7skipped)[2] <- "MYH7B_exon7skipped"

myh_wide <- myh_wide %>% merge(e7skipped)

g <- ggplot(myh_wide, aes(x = MYH7B_exon7skipped, y = log2(MYH7/MYH6)))
g + geom_point()


myh_wide$log2_MYH7_MYH6 <- log2(myh_wide$MYH7/myh_wide$MYH6)
ggscatter(myh_wide, x = "MYH7B_exon7skipped", y = "log2_MYH7_MYH6",
          color = "condition",
          shape = "condition",
          add = "reg.line",                                 # Add regression line
          conf.int = FALSE,                                # Add confidence interval
          add.params = list(color = "blue",
                            fill = "lightgray")
          )+
  stat_cor(method = "pearson", label.x = 4, label.y = 10)  # Add correlation coefficien


```


Let's look at the relationship between beta expression and MYH7B_skipped expression.

```{r}
# pdf("lncMYB7B_MYH7_e#C24A4Axpression.pdf")
g <- ggscatter(myh_wide, x = "MYH7B_exon7skipped", y = "MYH7",
          color = "condition",
          shape = "condition",
          add = "reg.line",   
          palette = c("#424242", "#752D2D","#C24A4A" ),
          conf.int = TRUE,                                
          add.params = list(color = "black",
                            fill = "lightgray"))+
  stat_cor(method = "pearson", label.x = 9, label.y = 5000) + 
  ylab("MHY7 tpm") + 
  xlab("MYH7B_exon7skipped tpm") + 
  ggtitle("lncMYH7B vs. MYH7 expression")
show(g)
# dev.off()

"../../results/markDuplicates/SRR7426784Aligned.sortedByCoord.out.markDups.bam"
```



# Sashimi plot of region

To actually runn differential transcript usage, the instructions are here: http://bioconductor.org/packages/release/workflows/vignettes/rnaseqDTU/inst/doc/rnaseqDTU.html


