---
title: "MYH7-MYH8 ratio"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
source("../util/_util.R")
source("../util/_setup.R")
```

```{r}
# This version has the exon skipped transcript appended.
# gtf <- rtracklayer::import("../../gencode.v32.annotation.gtf")
# genes <- gtf[which(gtf$type == "gene")]
# rtracklayer::export(genes, "gencode.v32.annotation.genes.gtf")
genes <- rtracklayer::import("gencode.v32.annotation.genes.gtf")


counts <- read.table("../../results/featureCounts/merged_gene_counts.txt", header = T)

g2s <- counts[,c(1,2)]
names(g2s) <- c("gene_id", "gene_name")


# Trim the sample names.
colnames(counts) <- sapply(colnames(counts), function(x) {
  unlist(strsplit(x, "Aligned"))[[1]]
})
samples <- read.csv("../../samplesheet.csv")

samples$condition <- sapply(samples$sample_title, function(x) {
  gsub('[[:digit:]]+', '', x)
})


# For some reason these samples doesn't exist
# SRR7426830
# SRR7426847
samples <- samples %>% filter(run_accession != "SRR7426830")
samples <- samples %>% filter(run_accession != "SRR7426847")

# Convert of tpm
fc <- fCountReader("../../results/featureCounts/gene_counts/",
                       samples$run_accession,
                       "Aligned.sortedByCoord.out_gene.featureCounts.txt")
tpm <- countConverter(fc, return = "TPM") %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  merge(g2s)
```

# MHY7 / MYH8 tpm plots

We may want to take the ratio of alpha to beta within each sample, maybe from the raw counts, maybe from some DEseq normalized count step, but for now, let's just look at the TPMs across groups.

```{r}

myh <- tpm %>% filter(gene_name %in% c("MYH6", "MYH7")) %>%
  gather(run_accession, tpm, 2:63) %>%
  merge(samples)

# 6 is alpha
# 7 is beta

g <- ggplot(myh, aes(x = gene_name, y = tpm))
g + geom_point() + 
  facet_wrap(~condition)
```

